React/Recipes/Datx Store Provider.md

```diff
@@ -0,0 +1,188 @@
+## datx
+
+> [datx.dev - A mobx data store](https://datx.dev/)
+>
+> DatX is an opinionated JS/TS data store. It features support for simple property definition, references to other models, and first-class TypeScript support.
+>
+> To find out more about DatX, check out the [official documentation](https://datx.dev/).
+
+To set up datx store, first, we need to install DatX dependency. To do this, follow [the official documentation](https://datx.dev/docs/getting-started/installation).
+
+After that, we can [create a collection](https://datx.dev/docs/getting-started/configuring-the-collection) and [models](https://datx.dev/docs/getting-started/defining-models).
+
+```ts
+import { Collection } from '@datx/core';
+
+export class Client extends Collection {
+  public static types = [];
+}
+```
+
+`types` variable should be a populated array with the model that your app uses. For the sake of this demonstration, it will be empty, but in the last section of this page, some examples will be shown.
+
+> To find out more about Datx, check out the [official documentation](https://datx.dev/).
+
+### JSON:API Client
+
+If you are working with a JSON:API, `datx` can handle that too - hello `datx-jsonapi`. This library provides multiple decorators that you can use to adapt collection and models to work with JSON:API. Read more about [JSON:API](https://jsonapi.org/) and [datx-jsonapi](https://datx.dev/docs/jsonapi/jsonapi-getting-started)
+
+### Extra
+
+If you want to follow along completely, here is a list of all needed dependencies:
+
+- [datx](https://datx.dev/)
+- [dequal](https://github.com/lukeed/dequal)
+- [mobx](https://mobx.js.org/README.html)
+- [swr](https://swr.vercel.app/)
+
+After datx store is initialized, we will create a datx context that will enable us to use DatX in our web application. For the context, we will need a provider - `DatxProvider`and a hook - `useHook`.
+
+## DatxProvider
+
+> For frontend network layer we will use [SWR](https://swr.vercel.app/) library created by Vercel.
+
+```tsx
+// DatxProvider.tsx
+
+import { createContext, FC } from 'react';
+import { SWRConfig } from 'swr';
+import { dequal } from 'dequal/lite';
+
+import { compare } from 'utils/compare';
+import { Client } from './Client';
+
+export const DatxContext = createContext<Client>(null);
+
+interface IDatxProviderProps {
+  client: Client;
+}
+
+export const DatxProvider: FC<IDatxProviderProps> = ({ client, children }) => (
+  <DatxContext.Provider value={client}>
+    <SWRConfig
+      value={{
+        compare: dequal,
```

Comment:
In terms of a core datx, is it necessary to have a custom compare function? Looking at the [dequal docs](https://github.com/lukeed/dequal#dequallite), it should be able to compare class instances. I'm correct or not?

Reply:
I'm not sure anymore, but we had some problem with default compare before on some project. 
@DarkoKukovec do you remember?

Also you are importing some custom `compare` function on line 51 which is not used anywhere. What's with that?

Reply:
Not sure, I'm not aware of any issues, but I wasn't that involved in this part, so I might have missed it.

Reply:
@DarkoKukovec I think you were involved in this conversation with @danijelbuhin or @mamihelj09 or someone. 
Problem happens when API returns new data and we already have previous data, SWR will try to compare previous value and new one. But as MobX is mutable, previous state in the SWR cache will also be updated and compare function will have the two identical objects and change won't be detected. 
I think something like that happened, it's because of MobX mutability.

Something like this:
```
const data = new MobxObservableObject({ name: 'test' }); // I forgot the syntax sorry 

// SWR cache
const cache = [{ key: 'some/url', value: data }];

// update happens
mutate(() => data.save());

// SWR compare will take the previous data in the cache and compare it with the new, and if there is change it will trigger component update
compare(cache[0].value, data);
```
As data is mutable observable object it will be updated in the SWR cache immediately after save is resolved, so new state and the old state will be the same...

Reply:
Yes, that's why the snapshots should be used - they should be immutable